src/test/java/org/junit/platform/engine/test/event/ExecutionEventRecorder.java

```diff
@@ -0,0 +1,175 @@
+
+/**
+ * {@link EngineExecutionListener} that records all events and makes them available to tests.
+ *
+ * @since 1.0
+ * @see ExecutionEvent
+ */
+public class ExecutionEventRecorder implements EngineExecutionListener {
+
+	public static List<ExecutionEvent> execute(TestEngine testEngine, EngineDiscoveryRequest discoveryRequest) {
+		TestDescriptor engineTestDescriptor = testEngine.discover(discoveryRequest,
+			UniqueId.forEngine(testEngine.getId()));
+		ExecutionEventRecorder listener = new ExecutionEventRecorder();
+		testEngine.execute(
+			new ExecutionRequest(engineTestDescriptor, listener, discoveryRequest.getConfigurationParameters()));
+		return listener.getExecutionEvents();
+	}
+
+	public final List<ExecutionEvent> executionEvents = new CopyOnWriteArrayList<>();
+
+	@Override
+	public void dynamicTestRegistered(TestDescriptor testDescriptor) {
+		addEvent(ExecutionEvent.dynamicTestRegistered(testDescriptor));
+	}
+
+	@Override
+	public void executionSkipped(TestDescriptor testDescriptor, String reason) {
+		addEvent(ExecutionEvent.executionSkipped(testDescriptor, reason));
+	}
+
+	@Override
+	public void executionStarted(TestDescriptor testDescriptor) {
+		addEvent(ExecutionEvent.executionStarted(testDescriptor));
+	}
+
+	@Override
+	public void executionFinished(TestDescriptor testDescriptor, TestExecutionResult result) {
+		addEvent(ExecutionEvent.executionFinished(testDescriptor, result));
+	}
+
+	@Override
+	public void reportingEntryPublished(TestDescriptor testDescriptor, ReportEntry entry) {
+		addEvent(ExecutionEvent.reportingEntryPublished(testDescriptor, entry));
+	}
+
+	public List<ExecutionEvent> getExecutionEvents() {
+		return executionEvents;
+	}
+
+	public Stream<ExecutionEvent> eventStream() {
+		return getExecutionEvents().stream();
+	}
+
+	public long getTestSkippedCount() {
+		return testEventsByType(SKIPPED).count();
+	}
+
+	public long getTestStartedCount() {
+		return testEventsByType(STARTED).count();
+	}
+
+	public long getReportingEntryPublishedCount() {
+		return testEventsByType(REPORTING_ENTRY_PUBLISHED).count();
+	}
+
+	public long getDynamicTestRegisteredCount() {
+		return testEventsByType(DYNAMIC_TEST_REGISTERED).count();
+	}
+
+	public long getTestFinishedCount() {
+		return testEventsByType(FINISHED).count();
+	}
+
+	public long getTestSuccessfulCount() {
+		return getTestFinishedCount(Status.SUCCESSFUL);
+	}
+
+	public long getTestAbortedCount() {
+		return getTestFinishedCount(Status.ABORTED);
+	}
+
+	public long getTestFailedCount() {
+		return getTestFinishedCount(Status.FAILED);
+	}
+
+	public long getContainerSkippedCount() {
+		return containerEventsByType(SKIPPED).count();
+	}
+
+	public long getContainerStartedCount() {
+		return containerEventsByType(STARTED).count();
+	}
+
+	public long getContainerFinishedCount() {
+		return containerEventsByType(FINISHED).count();
+	}
+
+	public List<ExecutionEvent> getSkippedTestEvents() {
+		return testEventsByType(Type.SKIPPED).collect(toList());
```

Comment:
It's not clear to me here if you want the list returned here to be mutable or immutable (the javadoc for `toList()` makes no guarantees on this, so the mutability of the list may change from JVM to JVM).

I'd strongly suggest changing it to either `.collect(toCollection(ArrayList::new))` or `.collect(collectingAndThen(toList(), Collections::unmodifiableList))`.